src/dsp_tools/analyse_xml_data/extract_links_from_XML.py

```diff
@@ -93,136 +103,239 @@ def _make_graph(
-    while g.num_nodes():
-        print(f"total number of nodes remaining: {g.num_nodes()}")
-        cycle = rx.digraph_find_cycle(g)  # type: ignore[attr-defined]  # pylint: disable=no-member
-        print("-" * 10)
-        print(f"cycle: {cycle}")
-        node = _find_cheapest_node(g, cycle, node_index_lookup)
-        source, targets = node
-        removed_nodes.append(UploadResource(g[source][0], targets))
-        g.remove_node(source)
-        removed_from_cycle += 1
-        print(f"removed link: {node}")
-        leaf_nodes = _remove_leaf_nodes(g, node_index_lookup)
-        removed_nodes.extend(leaf_nodes)
-    print("=" * 80)
-    print(f"removed links total: {removed_from_cycle}")
-    return removed_nodes
+    upload_order: list[str] = []
+    stash_lookup: dict[str, list[str]] = {}
+    leaf_nodes, node_indices = _remove_leaf_nodes(g, node_index_lookup, node_indices)
+    upload_order.extend(leaf_nodes)
+    stash_counter = 0
+    while node_indices:
+        cycle = list(rx.digraph_find_cycle(g))  # type: ignore[attr-defined]  # pylint: disable=no-member
+        links_to_remove = _find_cheapest_outgoing_links(g, cycle, edge_list)
+        stash_counter += len(links_to_remove)
+        links_to_stash = _remove_edges_get_removed_class_instances(
+            g=g,
+            edges_to_remove=links_to_remove,
+            edge_list=edge_list,
+            remaining_nodes=node_indices,
+        )
+        stash_lookup = _add_stash_to_lookup_dict(stash_lookup, links_to_stash)
+        leaf_nodes, node_indices = _remove_leaf_nodes(g, node_index_lookup, node_indices)
+        upload_order.extend(leaf_nodes)
+    return stash_lookup, upload_order, stash_counter
 
 
-def analyse_circles_in_data(xml_filepath: Path, tracer_output_file: str) -> list[UploadResource]:
+def analyse_circles_in_data(xml_filepath: Path, tracer_output_file: str, save_tracer: bool = False) -> None:
     """
     This function takes an XML filepath
     It analyzes how many and which links have to be removed
     so that all circular references are broken up.
+    This is for analysis purposes only.
 
     Args:
         xml_filepath: path to the file
         tracer_output_file: name of the file where the viztracer results should be saved
-
-    Returns:
-        The order in which the resources should be uploaded.
+        save_tracer: True if the output of the viztracer should be saved
     """
     start = datetime.now()
     print("=" * 80)
     tracer = VizTracer(
         minimize_memory=True,
-        ignore_c_function=True,
-        ignore_frozen=True,
-        include_files=["extract_links_from_XML.py", "models.py"],
+        max_stack_depth=3,
     )
     tracer.start()
     tree = etree.parse(xml_filepath)
     root = tree.getroot()
-    resptr_instances, xml_instances, all_resource_ids = _create_info_from_xml_for_graph(root)
+    root, resptr_instances, xml_instances, all_resource_ids = _create_info_from_xml_for_graph(root)
```

Comment:
Perhaps, `resptr_links` and `xml_links` would be better variable names

--> and in `_make_graph()`, too